applications/vanilla/library/class.categorycollection.php

```diff
@@ -0,0 +1,417 @@
+            $keys[] = $this->cacheKey(self::$CACHE_CATEGORY, $id);
+        }
+        if (!empty($keys)) {
+            $cacheCategories = $this->cache->get($keys);
+            foreach ($cacheCategories as $key => $category) {
+                $this->categories[(int)$category['CategoryID']] = $category;
+                $this->categorySlugs[$category['UrlCode']] = (int)$category['CategoryID'];
+
+                $categories[(int)$category['CategoryID']] = $category;
+            }
+        }
+
+        // Look in the database.
+        $dbCategoryIDs = [];
+        foreach ($categories as $id => $row) {
+            if (!$row) {
+                $dbCategoryIDs[] = $id;
+            }
+        }
+        $dbCategories = $this->sql->getWhere('Category', ['CategoryID' => $dbCategoryIDs])->resultArray();
+        foreach ($dbCategories as &$category) {
+            $this->calculate($category);
+
+            $this->cache->store(
+                $this->cacheKey(self::$CACHE_CATEGORY, $category['CategoryID']),
+                $category
+            );
+            $this->cache->store(
+                $this->cacheKey(self::$CACHE_CATEGORY_SLUG, $category['UrlCode']),
+                (int)$category['CategoryID']
+            );
+
+            $this->categories[(int)$category['CategoryID']] = $category;
+            $this->categorySlugs[$category['UrlCode']] = (int)$category['CategoryID'];
+
+            $categories[(int)$category['CategoryID']] = $category;
+        }
+
+        return $categories;
+    }
+
+    /**
+     * Insert a new category, handling its tree properties and caching.
+     *
+     * This method is currently only to to be used in a support role.
+     *
+     * @param array $category The new category.
+     */
+    public function insert(array $category) {
+        $category += [
+            'DateInserted' => Gdn_Format::toDateTime(),
+            'InsertUserID' => 1,
+            'ParentCategoryID' => -1,
+        ];
+
+        // Filter out fields that aren't in the table.
+        $category = array_intersect_key($category, $this->getSchema()->fields());
+        $categoryID = $this->sql->insert('Category', $category);
+
+        if ($categoryID) {
+            // Update my parent's count.
+            $this->sql->put('Category', ['CountCategories+' => 1], ['CategoryID' => $category['ParentCategoryID']]);
+
+            $this->refreshCache($category['CategoryID']);
+            $this->refreshCache($category['ParentCategoryID']);
+        }
+
+        return $categoryID;
+    }
+
+    /**
+     * Update an existing category, handling its tree properties and caching.
+     *
+     * @param array $category The category to update.
+     * @return bool Returns **true** if the category updated or **false** otherwise.
+     */
+    public function update(array $category) {
+        if (empty($category['CategoryID'])) {
+            throw new Gdn_UserException("Category ID is required.");
+        }
+
+        $category += [
```

Comment:
I don't think that the update function is used yet but this would not update the `DateUpdated` field on a category that already has it. (Same for the UpdateUserID)